cpp/src/slice2cpp/Gen.cpp

```diff
@@ -898,289 +874,184 @@ Slice::Gen::writeExtraHeaders(IceInternal::Output& out)
+        {
+            const string& argument = meta->arguments();
+            if (argument != "string" && argument != "wstring")
             {
-                continue;
+                return "invalid argument '" + argument + "' supplied to 'cpp:type' metadata in this context.";
             }
+            return nullopt;
         }
-        if (dynamic_pointer_cast<Sequence>(cont))
+
+        // Now that we checked all the containers we only need to worry about 'cpp:type's that've been applied to types.
+        // If we're looking at an operation, parameter, or data member, we want to be checking its type instead.
+        SyntaxTreeBasePtr appliedTo = p;
+        if (auto op = dynamic_pointer_cast<Operation>(appliedTo))
         {
-            if ((directive == "cpp:type" || directive == "cpp:view-type") && !arguments.empty())
+            if (const auto returnType = op->returnType())
             {
-                continue;
+                appliedTo = returnType;
             }
-            if (directive == "cpp:array" && arguments.empty())
-            {
-                continue;
-            }
-        }
-        if (dynamic_pointer_cast<Dictionary>(cont))
-        {
-            if ((directive == "cpp:type" || directive == "cpp:view-type") && !arguments.empty())
+            else
             {
-                continue;
+                return "'cpp:type' metadata cannot be applied operations with void return type";
             }
         }
-        if (dynamic_pointer_cast<Exception>(cont) && directive == "cpp:ice_print" && arguments.empty())
+        else if (auto param = dynamic_pointer_cast<ParamDecl>(appliedTo))
         {
-            continue;
+            appliedTo = param->type();
         }
-        if (dynamic_pointer_cast<Enum>(cont) && directive == "cpp:unscoped" && arguments.empty())
+        else if (auto dm = dynamic_pointer_cast<DataMember>(appliedTo))
         {
-            continue;
+            appliedTo = dm->type();
         }
 
+        // Finally, we check the metadata against the type it's being applied on.
+        if (auto builtin = dynamic_pointer_cast<Builtin>(appliedTo); builtin && builtin->kind() == Builtin::KindString)
         {
-            ClassDeclPtr cl = dynamic_pointer_cast<ClassDecl>(cont);
-            if (cl && directive == "cpp:type" && !arguments.empty())
+            const string& argument = meta->arguments();
+            if (argument != "string" && argument != "wstring")
             {
-                continue;
+                return "invalid argument '" + argument + "' supplied to 'cpp:type' metadata in this context.";
             }
+            return nullopt;
         }
+        else if (dynamic_pointer_cast<Sequence>(appliedTo) || dynamic_pointer_cast<Dictionary>(appliedTo) || dynamic_pointer_cast<ClassDecl>(appliedTo))
+        {
+            return nullopt; // TODO: I see no reason to support 'cpp:type' on class declarations.
```

Comment:
I really wanted to avoid changing any behavior in this PR, since it also changes alot of other stuff.
After this PR is merged, I'll open a separate PR to remove this, and any of the other weird metadata things that are easier to see now.